sdk/core/azure_core/src/cloud.rs

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+//! Azure cloud configuration.
+
+use std::{any::TypeId, collections::HashMap};
+
+/// Configurations for different Azure clouds.
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+#[non_exhaustive]
+pub enum CloudConfiguration {
+    /// Azure Public Cloud
+    #[default]
+    AzurePublic,
+
+    /// Azure Government
+    AzureGovernment,
+
+    /// Azure in China
+    AzureChina,
+
+    /// A custom cloud.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # mod azure_service_module {
+    /// #   pub struct Audience;
+    /// # }
+    ///
+    /// use azure_core::{
+    ///     cloud::{Audiences, CloudConfiguration, CustomConfiguration},
+    ///     http::ClientOptions,
+    /// };
+    ///
+    /// let mut custom = CustomConfiguration::default();
+    /// custom.audiences = Audiences::new().with::<azure_service_module::Audience>("https://service.mycloud.local".to_string());
+    /// custom.authority_host = "https://login.mycloud.local".to_string();
+    /// let cloud: CloudConfiguration = custom.into();
+    /// ```
+    Custom(CustomConfiguration),
+}
+
+/// Configuration for a custom cloud.
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+#[non_exhaustive]
+pub struct CustomConfiguration {
+    /// Base URL for authentication, for example "https://login.microsoftonline.com"
+    pub authority_host: String,
+
+    /// Map of SDK modules to their Entra ID audiences.
+    pub audiences: Audiences,
+}
+
+impl From<CustomConfiguration> for CloudConfiguration {
+    fn from(config: CustomConfiguration) -> Self {
+        Self::Custom(config)
+    }
+}
+
+/// Collection of audiences for an Azure cloud's services
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct Audiences(HashMap<TypeId, String>);
+
+impl Audiences {
+    /// Create an empty `Audiences` map.
+    pub fn new() -> Self {
+        Self(HashMap::new())
+    }
+
+    /// Get a module's audience.
+    pub fn audience<T: 'static>(&self) -> Option<&str> {
+        self.0.get(&TypeId::of::<T>()).map(|s| s.as_str())
+    }
+
+    /// Insert or replace an audience and return `Self` to allow chaining.
+    pub fn with<T: 'static>(mut self, audience: String) -> Self {
```

Comment:
Probably should add an `insert<T>(&mut self)` as well.

Reply:
Also, this only sets the `audience` so should be `with_audience` per guidelines. Alternatively, keep them both generic and just have `get` and `with`, which might make more sense. The idea is to have something of the name - the getter and setter, if you will - linking the two obviously.

Reply:
My reasoning was that `_audience` is implicit because this is `Audiences::with()`, the collection holds non-compound values (Strings), and client modules will probably name their key type `Audience`. So, I like `get` and `with`